### DISCUSSION TITLE
Github pages and CORS

### DISCUSSION THREAD

**Discussion by norricorp** [community/community#22399] [Other Feature Feedback, Questions, & Ideas]
<p>Hello,<br>
my set up is an app on github pages (GH) that calls a local rest api server to get data.<br>
I can access the data from outside my office network via curl. The data is behind https.</p>
<p>But when the page from github pages tries to access the data, the browser console shows<br>
<code>Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://xx.xx.xx.xx:3000/users/login. (Reason: CORS request did not succeed).</code></p>
<p>The server has CORS enabled (it is the loopback 4 framework that has CORS enabled by default). Is there anything I need to do in the app under GH?</p>

**Comment by yoannchaudet** [comment]
<p>Hello,</p>
<p>Does your Pages site has its visibility set to private?</p>
<p>Pages allows CORS (<code>access-control-allow-origin</code> header is set to <code>*</code>) for public sites only.</p>
<p>This is the current (sensitive) behavior in place until we can support custom headers.</p>

**Comment by norricorp** [comment]
<p>No the site is public.<br>
I wonder if I need to add the mode attribute to fetch except that the default is “cors” so perhaps not.</p>

**Comment by yoannchaudet** [comment]
<p>Only thing I can think of is extra restrictions your browser would put on. Say if <code>xx.xx.xx.xx</code> where an IP address that is considered local (e.g. 192.168.x.x).</p>

**Comment by norricorp** [comment]
<p>The server VM is local but within the code on github, it points to my ISP allocated IP address and the port forwarding gets it to the VM.</p>

**Comment by norricorp** [comment]
<p>I have added “mode” (cors) to the fetch request along with a header of “Access-Control-Allow-Origin” (the IP address of my ISP connection).<br>
The same result but then I realised that I had not opened my port forwarding this time. So I opened it and got a slightly different error message.</p>
<p><code>Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://88.111.150.77:3000/users/login. (Reason: CORS header 'Access-Control-Allow-Origin' missing).</code></p>
<p>Now this is the first time I have had a reason that mentioned Access-Control-Allow-Origin. It is also the first time that I specifically added Access-Control-Allow-Origin to the headers in the fetch request.</p>
<p>Which is all very confusing.</p>

**Comment by airtower-luna** [comment]
<p><code>Access-Control-Allow-Origin</code> is a response header, setting it in the request shouldn’t have any effect. Your local server would need to set to allow the browser to provide its response to your GH Pages site (which has a different origin). See <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin" class="inline-onebox">Access-Control-Allow-Origin - HTTP | MDN</a>.</p>

**Comment by norricorp** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a>, thank you for this.<br>
I have read that and whenever I think I understand CORS, it comes back to bite me. Perhaps I just have a very non-standard setup.<br>
Anyway, I have changed the loopback 4 api server to return the header. But same original error now. Looking at the headers, there is no reponse header info.<br>
Just to add, I added logging to the rest api server and the request is not being reached and that is where I added the header.</p>

**Reply by GrantHough** [reply]
bro felt every time i fix a cors bug it just remanifests at some point

**Comment by norricorp** [comment, marked as answer]
<p>So I solved this. Nothing to do with github pages or svelte.<br>
I was using the cors origin value in the rest server as the url of the server. It needs to be the <a href="http://github.io" rel="noopener nofollow ugc">github.io</a> url.</p>

**Comment by Sam-cudo** [comment]
<p>Can you be more precise about how you solved this problem</p>

**Comment by norricorp** [comment]
<p>Hello <a class="mention" href="/sam-cudo">@Sam-cudo</a>,<br>
my backend on this is loopback 4 - it is a node framework from IBM.<br>
An app in loopback has an index.js file which allows you to set the CORS configuration .<br>
So I added<br>
cors: {<br>
origin: [‘<a href="https://xxxxxx.github.io" rel="noopener nofollow ugc">https://xxxxxx.github.io</a>’, ‘<a href="http://localhost:5000" rel="noopener nofollow ugc">http://localhost:5000</a>’],<br>
methods: ‘GET,HEAD,PUT,PATCH,DELETE’,<br>
},</p>
<p>I have two entries for CORS because the frontend code is either running on a local svelte server (localhost/PC) or on github.<br>
There is more on loopback at <a href="https://loopback.io/doc/en/lb4/Customizing-server-configuration.html#customize-cors" class="inline-onebox" rel="noopener nofollow ugc">Customizing Server configuration | LoopBack Documentation</a></p>
<p>I hope this helps. The key thing for me was realising that the backend had to know where to expect requests to come from; in my case either <a href="http://github.io" rel="noopener nofollow ugc">github.io</a> or my local svelte server (or apache or nginx or whatever)</p>

**Comment by dotku** [comment]
Any update about it?
I still having problem on CORS on codepen: https://codepen.io/pen/?editors=0010
Let me know if there is solution.

**Comment by IntangibleMatter** [comment]
I'm running into issues with this too- I'm using a custom CNAME for my pages site (intangiblematter.net) and I'm getting info about the request being blocked. I'm trying to request from the Bluesky API. Is it possible that there's some thing in requesting that I'm missing?

For reference, this is my request
```js
// searchParams is just a URLSearchParams object for the request.
const searchResponse = await fetch(
	`https://public.api.bsky.app/xrpc/app.bsky.feed.searchPosts?${searchParams}`,
	{
		headers: {
			Accept: "application/json",
			// as you can see I've been trying various things and they haven't been working
			origin: "https://intangiblematter.github.io",
			Referer: "https://intangiblematter.github.io",
		},
	},
);
```
